Add vmaas_sync

blayson opened the pull request.

jdobes commented on base/database/setup.go:
```diff
@@ -1,5 +1,74 @@
+}
+
+func loadDbConfig() *PostgreSQLConfig {
+	config := PostgreSQLConfig{
```
Seems good for now, but we should also get values from clowder bit later.

blayson replied:
agreed, need to investigate how to do it in Go :D

epapbak commented on Dockerfile:
```diff
@@ -20,6 +20,7 @@ ADD manager                     /vuln4shift/manager
 ADD pyxis                       /vuln4shift/pyxis
 ADD scripts                     /vuln4shift/scripts
 ADD test                        /vuln4shift/test
+ADD vmsync                        /vuln4shift/vmsync
```
please align =)

blayson replied:
Thats quite interesting, fixed :)

epapbak commented on base/logging/logger.go:
```diff
@@ -27,7 +27,7 @@ func ParseLoglevel(level string) (logrus.Level, error) {
 		return logrus.FatalLevel, nil
 	default:
-		return logrus.InfoLevel, errors.New("Invalid loglevel given")
+		return logrus.InfoLevel, errors.New("invalid loglevel given")
```
I'm curious, why this change?

blayson replied:
Error string should not be capitalized or end with punctuation mark 
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

epapbak commented on base/utils/database.go:
```diff
@@ -4,27 +4,17 @@ import (
-	dbName := getDbVar("POSTGRES_DB")
-	dbUser := getDbVar("POSTGRES_USER")
-	dbPassword := getDbVar("POSTGRES_PASSWORD")
+	dbHost := GetEnv("POSTGRES_HOST", "localhost")
```
If we are introducing default values here, it should be `vuln4shift_database` and `5432` right?

blayson replied:
oh, I have tested it locally and forgot to change, will fix

epapbak commented on go.mod:
```diff
@@ -1,12 +1,13 @@
 module app
 
-go 1.16
+go 1.18
```
since we are not using anything specific from this version, what do you think about leaving this change for another PR?

jdobes replied:
agree, also we might need to wait until we're able to compile with 1.18, https://catalog.redhat.com/software/containers/ubi8/go-toolset/5ce8713aac3db925c03774d1?container-tabs=overview is still on 1.16

jdobes replied:
- In RHEL/Fedora/UBI there are no 1.18 RPM packages yet, we'd need custom UBI-based build image and install go 1.18 binaries manually. Or wait on Fedora 36, which is going out later this month, should contain golang 1.18 packages and base the build image on Fedora
- Or try to replace UBI-golang build image with upstream https://hub.docker.com/_/golang

jdobes replied:
The second option could be quite good, unless there will be any problem/incompatibility

jdobes replied:
And if we switch to 1.18 let's do the change in separate PR, which can be merged before this one

blayson replied:
seems, like simple replace of builder image works

blayson replied:
@jdobes #23

epapbak reviewed (commented):
some minor stuff

jdobes commented on go.mod:
```diff
@@ -15,3 +16,49 @@ require (
 	gorm.io/gorm v1.23.2
 )
+
+require (
```
Where does these indirect deps come from? Are there all deps from go.sum? Is this generated? Do these indirect deps need to be updated manually?

blayson replied:
yeah this were generated by `go mod tidy` command

jdobes replied:
found the reason - https://stackoverflow.com/questions/69432877/why-there-are-two-require-blocks-in-go-mod-since-go-1-17?noredirect=1&lq=1

previously, the go.mod was generated on Fedora, where is still golang 1.16, 1.18 will be in Fedora 36 later this month

epapbak commented on vmsync/api.go:
```diff
@@ -0,0 +1,91 @@
+
+		statusCode, err := client.RetryRequest(http.MethodPost, VmaasCvesURL, &vmaasRequest, &vmaasResponse)
+		if err != nil {
+			logger.Warningf("Request %s %s failed: statusCode=%d, err=%s", http.MethodPost, VmaasCvesURL, statusCode, err)
```
I would use logger.Errorf since when this happens, we stop the execution. WDYT?

jdobes replied:
I'd keep Warning, pyxis gatherer api.go is using the same. If needed it can be then changed on both places at once.

epapbak commented on vmsync/vmsync.go:
```diff
@@ -0,0 +1,163 @@
+	if err != nil {
+		logger.Fatalf("Unable to get CVEs from VMaaS: %s", err)
+	}
+	logger.Infof("CVEs in VMaaS: %d", len(apiCveMap))
```
It basically says the same as the log[ just before](https://github.com/RedHatInsights/vuln4shift-backend/pull/5/files#diff-f03f754ac47dfa235847556581a3690e853ae0fc4e8659ffea4a01645cd45df7R87), in `getAPICves`

blayson replied:
deleted

epapbak commented on vmsync/vmsync.go:
```diff
@@ -0,0 +1,163 @@
+	logger.Infof("CVEs to delete: %d", len(toDeleteCves))
+
+	if err = syncCves(toSyncCves, toDeleteCves); err != nil {
+		logger.Infof("Error during syncing CVEs into database: %s", err)
```
`logger.Errorf`, maybe?

jdobes replied:
Perhaps Fatalf, because we don't want to continue from here and print the success msg below

epapbak reviewed (commented):
Small comments. Aside from that, LGTM

jdobes commented on base/models/utils.go:
```diff
@@ -3,10 +3,13 @@ package models
 func GetGormConnection(dsn string) (*gorm.DB, error) {
-	db, err := gorm.Open(postgres.Open(dsn))
+	db, err := gorm.Open(postgres.Open(dsn), &gorm.Config{
+		Logger: logger.Default.LogMode(logger.Error),
```
What is this logging when added? Why logger.Error?

blayson replied:
That was done to debug the queries, but then I have leave that in Error mode, as if I got it correctly it is a default mode.

jdobes replied:
Ok, thanks

jdobes commented on conf/vmsync.env:
```diff
@@ -0,0 +1,6 @@
+LOGGING_LEVEL=INFO
+API_RETRIES=3
+PAGE_SIZE=5000
+BASE_URL=http://localhost:8080/api/v3/cves
```
I'd just set `https://console.redhat.com/api/vmaas/v3/cves` here, not relying on local vmaas

yungbender replied:
also newline pls

jdobes commented on vmsync/api.go:
```diff
@@ -0,0 +1,91 @@
+			return cveMap, err
+		}
+
+		cveList, err := json.Marshal(&vmaasResponse.CveList)
```
Why marshall and then unmarshall? Why not just fill `cveMap` in `for`?

blayson replied:
changed

jdobes commented on vmsync/vmsync.go:
```diff
@@ -0,0 +1,163 @@
+			severity = models.NotSet
+		}
+
+		cvss2Score, _ := strconv.ParseFloat(apiCve.Cvss2Score, 32)
```
Better check the err

blayson replied:
Yep, done.
If the error I'm setting `cvss2/3Score` to 0.0

jdobes commented on vmsync/vmsync.go:
```diff
@@ -0,0 +1,163 @@
+	return tx.Commit().Error
+}
+
+func deleteCves(toDeleteCves []models.Cve, tx *gorm.DB) error {
```
Let's not delete CVEs yet.

- It shouldn't work this way - it shouldn't forcibly delete all deps from other tables (vuln-engine deletes only CVEs which are not on any systems)
- keep the function here, just don't call it for now
- there will be changes needed on both this side and on pyxis gatherer side to fix concurrency issues on the cve table (deadlocks during insert and update) and the CVE cache which is created in both components (and will contain invalid data when some CVE is deleted)

blayson replied:
Got it.

yungbender commented on .gitignore:
```diff
@@ -15,3 +15,5 @@
 
 .idea/
+
+vendor/
```
newline pls

yungbender commented on dbadmin/migrations/004_grant_insert_update_vmaas_gatherer_down.sql:
```diff
@@ -0,0 +1,4 @@
+REVOKE INSERT, UPDATE, DELETE ON cve IN SCHEMA public FROM vmaas_gatherer;
+REVOKE DELETE ON account_cve_cache IN SCHEMA public FROM vmaas_gatherer;
+REVOKE DELETE ON image_cve IN SCHEMA public FROMvmaas_gatherer;
+REVOKE DELETE ON cluster_cve_cache IN SCHEMA public FROM vmaas_gatherer;
```
newline pls

yungbender commented on dbadmin/migrations/004_grant_insert_update_vmaas_gatherer_up.sql:
```diff
@@ -0,0 +1,4 @@
+GRANT SELECT, INSERT, UPDATE, DELETE ON cve TO vmaas_gatherer;
+GRANT SELECT, DELETE ON account_cve_cache TO vmaas_gatherer;
+GRANT SELECT, DELETE ON image_cve TO vmaas_gatherer;
+GRANT SELECT, DELETE ON cluster_cve_cache TO vmaas_gatherer;
```
newline pls

yungbender commented:
@blayson I have found a problem when running it, the sql migration scripts are named `*_down.sql` `*_up.sql` , this wont work, it needs to be named `*.down.sql` and `*.up.sql`

blayson commented:
> @blayson I have found a problem when running it, the sql migration scripts are named `*_down.sql` `*_up.sql` , this wont work with the go-migrate, it needs to be named `*.down.sql` and `*.up.sql`. After I changed it locally it works and syncs correctly 👍

Yeah, thanks, renamed

jdobes commented:
removing these checks from .golangci.yaml, is it intentional? If yes, why?

blayson commented:
> removing these checks from .golangci.yaml, is it intentional? If yes, why?

There was a warning that this checks are disabled in go1.18. So, I have decided to remove it from golangci.yaml as it has no effect.
